### PR TITLE
Secret storage interface

### PR DESCRIPTION
**Pull request by ilackarms**
@yuval-k and @axhixh PTAL; `glooctl` will be able to leverage this

### CODE REVIEW

**Comment by ilackarms** [comment]
this pr adds an interface to `storage` that can be easily consumed by different components of gloo who wish to read/write secrets to a supported storage layer. follows similar conventions to our config store client

**Comment by yuval-k** [comment]
I noticed you removed my .keep file in gloo_config the nats example.
The reason I put them there is because otherwise docker will create those folders with root permissions. were they causing you an issue?

File: pkg/storage/dependencies/kube/secretorage_test.go [hunk 1]
```diff
@@ -0,0 +1,206 @@
...
+func toStringMap(m map[string][]byte) map[string]string {
+	sm := make(map[string]string)
+	for k, v := range m {
+		sm[k] = string(v)
+	}
+	return sm
+}
+
+var _ = FDescribe("Secret Storage Client", func() {
+	if os.Getenv("RUN_KUBE_TESTS") != "1" {
```
Comment (yuval-k) [inline pkg/storage/dependencies/kube/secretorage_test.go]: suggestion: consider moving this file to e2e folder as it is not self contained (as depends on kube).

File: pkg/storage/dependencies/vault/secretorage_test.go [hunk 2]
```diff
@@ -0,0 +1,188 @@
...
+	"time"
+
+	"github.com/hashicorp/vault/api"
+	"github.com/solo-io/gloo/pkg/log"
+	"github.com/solo-io/gloo/pkg/storage/dependencies"
+	. "github.com/solo-io/gloo/test/helpers"
+)
+
+var _ = Describe("Secret Storage Client", func() {
+	if os.Getenv("RUN_VAULT_TESTS") != "1" {
```
Comment (yuval-k) [inline pkg/storage/dependencies/vault/secretorage_test.go]: consider moving this file to e2e

**Comment by yuval-k** [comment]
@ilackarms few small comments

File: pkg/storage/dependencies/vault/secretorage_test.go (hunk 2 above)
Comment (ilackarms) [reply pkg/storage/dependencies/vault/secretorage_test.go]: it's an integration test for the vault secrets client

we currently have a pattern of integration tests living next to the units they test

**Comment by ilackarms** [comment]
ah, didn't realize there was a good reason for the .keep
i'll put them back in

**Comment by ilackarms** [comment]
@yuval-k as far as the `.keep` files, the reason they're deleted is because the `.gitignore` ignores files with `_gloo_config`. do we want to remove that from the gitignore?

**Comment by ilackarms** [comment]
sure

**Pull request merged**